### DISCUSSION TITLE
openai gpt-3.5 adding a strange end string to translations

### DISCUSSION THREAD

**Discussion by JohnRDOrazio** [WeblateOrg/weblate#11084] [Q&A] [answered]
I updated my self-hosted Weblate instance today and decided to test out the new OpenAI integration. It's working pretty great, I have succeeded in instructing it to preserve XML tags and it has obeyed 😄 

However I'm finding some strange strings added onto the end of the translated string in a number of cases:

```
 
==WEBLATE_PART==
```

Basically it's adding a carriage return or newline character and then `==WEBLATE_PART==`. Before creating an issue, does any know what could be happening here? And if there's a simple fix to prevent it from happening?

In the configuration page I have basically used the suggested phrase for configuration with a slight tweak just to contextualize in the specific business area for these translations:
`You are a highly skilled translation assistant [for such and such an industry], adept at translating text between languages with precision and nuance. You always reply with translated string only. You do not include transliteration. And you keep XML tags in place.`

Adding that last specification ` And you keep XML tags in place.` worked quite well in preserving formatting tags included in these strings. But even before adding that I was already getting these strange strings, sometimes even preprended. Here's an example, I have a simple single word (title of a paragraph in a book) that needs to get translated, here is the original string in the source JSON file:

```js
	"Story_I": {
		"PSR_0": {
			"CSR_0": {
				"Content_0": "Introduzione"
			}
		},
```

As you can see, there are no newline characters or carriage return characters anywhere in the original string. However this is what I'm seeing as a suggestion from OpenAI GPT-3.5:
![Screenshot weblate openai](https://github.com/WeblateOrg/weblate/assets/231298/ec0bb8ea-4399-4bae-8e29-9bce4dd5de93)

**Comment by nijel** [comment, marked as answer]
This string is used by Weblate to separate strings when translating more of them. 2cf0ae88408c59dd476fcc82ea48c30970d5cd24 should improve it a bit.

**Comment by JohnRDOrazio** [comment]
I am also seeing this in an OpenAI translated string: `<trans-unit id="1">` (Weblate 5.4.3)
 ![image](https://github.com/WeblateOrg/weblate/assets/231298/b48b4eb5-cd4a-417c-8b62-b3792f26ea46)

**Reply by nijel** [reply]
That's strange, this is not anything Weblate uses internally, but it looks like an XLIFF snippet. Can you check your glossary for similarly looking strings? Besides that, it might be just OpenAI deciding this belongs to the translation…

**Reply by JohnRDOrazio** [reply]
Perhaps other than instructing OpenAI to "preserve XML tags", I should also tell it to "not add any XML tags".

**Reply by JohnRDOrazio** [reply]
That actually seems to have helped, I specified in the OpenAI configuration:

> You keep XML tags in place, but you do not add any XML tags that are not in the original string.

And here is the result:

![image](https://github.com/WeblateOrg/weblate/assets/231298/bc3eda6b-9ad1-4303-bc1e-eab014aa8a84)

Now just waiting for the `==WEBLATE_PART==` fix to be pulled into an upcoming update...

**Comment by JohnRDOrazio** [comment]
After updating to weblate 5.5.5 I am no longer seeing `==WEBLATE_PART==` in Open AI suggested translations!

**Reply by nijel** [reply]
Yes, 2cf0ae88408c59dd476fcc82ea48c30970d5cd24 should have addressed that.
